typescript.md

```diff
@@ -0,0 +1,283 @@
+# Guidelines for typing in TypeScript
+
+## Where to store types specific to a repository
+
+When types are shared between modules and used sparsely, create a `src/types.ts` file
+as per TypeScript's conventions, where you will pile up all the common types.
+
+Presumably you will also build a declaration at `lib/types.d.ts`.
+If your repository hosts a library, then it will be useful to have those types easily imported into parent modules.
+Consider adding `types.d.ts` in the root of your repository with the content `export * from './lib/types.d.ts';`,
+since this will make parent modules be able to write `import {FunkyType} from 'funky-module/types';`.
+
+
+## Casting and guards
+
+At times, you may need to clarify to typescript that a variable is certainly of a specific type within a block.
+
+```typescript
+let fn = function(a: boolean | string): boolean {
+  if ((a as string).match) {
```

Comment:
I found this example/pattern confusing. It took me a while just to understand the intention of the first check (perhaps because my first thought was that the match function did some magical thing using regexps to see if the type of a matched 'string'). But anyway, I don't think I like this pattern. Afaics it does not check if the variable is of a certain type, just that it has a specific function. Using this pattern here would not work:
```ts
let foo = [ 2, 3];

if ((foo as string).length) {
  console.log(`foo is a string of length ${foo.length);
}
```